src/map/pc.c

```diff
@@ -4392,43 +4392,45 @@ int pc_payzeny(struct map_session_data *sd,int zeny, enum e_log_pick_type type,
 /*==========================================
  * Cash Shop
  *------------------------------------------*/
-
+//Changed Kafrapoints calculation. [Normynator]
 int pc_paycash(struct map_session_data *sd, int price, int points)
 {
 	int cash;
-	nullpo_retr(-1,sd);
+	int mempoints;
+	nullpo_retr(-1, sd);
 
-	points = cap_value(points,-MAX_ZENY,MAX_ZENY); //prevent command UB
-	if( price < 0 || points < 0 )
-	{
+	points = cap_value(points, -MAX_ZENY, MAX_ZENY); //prevent command UB
+	if (price < 0 || points < 0) {
 		ShowError("pc_paycash: Paying negative points (price=%d, points=%d, account_id=%d, char_id=%d).\n", price, points, sd->status.account_id, sd->status.char_id);
 		return -2;
 	}
 
-	if( points > price )
-	{
+	if (points > price) {
 		ShowWarning("pc_paycash: More kafra points provided than needed (price=%d, points=%d, account_id=%d, char_id=%d).\n", price, points, sd->status.account_id, sd->status.char_id);
-		points = price;
+		points = points - price;
+		mempoints = price;
+		cash = 0;
+	} else {
+		cash = price - points;
+		mempoints = points;
+		points = 0;
 	}
 
-	cash = price-points;
-
-	if( sd->cashPoints < cash || sd->kafraPoints < points )
-	{
+	if (sd->cashPoints < cash || sd->kafraPoints < points) {
```

Comment:
At this Line:
sd->kafraPoints < points should be sd->kafraPoints < mempoints

going to change that, don't merge. Don't know if it is exploitable at the current state.
@Smokexyz